lib/srv/db/watcher.go

```diff
@@ -188,13 +199,129 @@ func (s *Server) matcher(resource types.ResourceWithLabels) bool {
 		return false
 	}
 
-	// In the case of CloudOrigin CloudHosted resources the matchers should be skipped.
-	if cloudOrigin(resource) && database.IsCloudHosted() {
+	// In the case of databases discovered by this database server, matchers
+	// should be skipped.
+	if s.monitoredDatabases.isCloud(database) {
 		return true // Cloud fetchers return only matching databases.
 	}
+
+	// Database resources created via CLI, API, or discovery service are
+	// filtered by resource matchers.
 	return services.MatchResourceLabels(s.cfg.ResourceMatchers, database)
 }
 
-func cloudOrigin(r types.ResourceWithLabels) bool {
-	return r.Origin() == types.OriginCloud
+type discoveryResourceChecker interface {
+	check(ctx context.Context, database types.Database)
+}
+
+type discoveryResourceCheckerImpl struct {
```

Comment:
nit: Could we drop the Imp suffix and name it something like `cloudValidCredentailsChecker` ?